File: filesystemstorage/src/main/java/com/ithit/webdav/samples/fsstorageservlet/extendedattributes/OSXExtendedAttribute.java
package com.ithit.webdav.samples.fsstorageservlet.extendedattributes;

import com.ithit.webdav.server.exceptions.ServerException;
import com.sun.jna.platform.mac.XAttrUtil;

/**
 * OS X extended attribute support using native API.
 */
class OSXExtendedAttribute implements ExtendedAttribute {

    /**
     * {@inheritDoc}
     */
    @Override
    public void setExtendedAttribute(String path, String attribName, String attribValue) throws ServerException {
        int result = XAttrUtil.setXAttr(path, attribName, attribValue);
        if (result == -1) {
            throw new ServerException(
                    String.format("Writing attribute '%s' with value '%s' to file '%s' failed.", attribName, attribValue, path));
        }
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public String getExtendedAttribute(String path, String attribName) throws ServerException {
        try {
            return XAttrUtil.getXAttr(path, attribName);
        } catch (Exception e) {
            throw new ServerException(
                    String.format("Reading attribute '%s' from file '%s' failed.", attribName, path));
        }
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public void deleteExtendedAttribute(String path, String attribName) throws ServerException {
        int result = XAttrUtil.removeXAttr(path, attribName);
        if (result == -1) {
            throw new ServerException(
                    String.format("Removing attribute '%s' from file '%s' failed.", attribName, path));
        }
    }
}

File: Kotlin/kotlinfs/src/main/kotlin/com/ithit/webdav/samples/kotlinfs/extendedattributes/OSXExtendedAttribute.kt
package com.ithit.webdav.samples.kotlinfs.extendedattributes

import com.ithit.webdav.server.exceptions.ServerException
import com.sun.jna.platform.mac.XAttrUtil

/**
 * OS X extended attribute support using native API.
 */
internal class OSXExtendedAttribute : ExtendedAttribute {

    /**
     * {@inheritDoc}
     */
    @Throws(ServerException::class)
    override fun setExtendedAttribute(path: String, attribName: String, attribValue: String) {
        val result = XAttrUtil.setXAttr(path, attribName, attribValue)
        if (result == -1) {
            throw ServerException(
                    String.format("Writing attribute '%s' with value '%s' to file '%s' failed.", attribName, attribValue, path))
        }
    }

    /**
     * {@inheritDoc}
     */
    @Throws(ServerException::class)
    override fun getExtendedAttribute(path: String, attribName: String): String {
        try {
            return XAttrUtil.getXAttr(path, attribName)
        } catch (e: Exception) {
            throw ServerException(
                    String.format("Reading attribute '%s' from file '%s' failed.", attribName, path))
        }

    }

    /**
     * {@inheritDoc}
     */
    @Throws(ServerException::class)
    override fun deleteExtendedAttribute(path: String, attribName: String) {
        val result = XAttrUtil.removeXAttr(path, attribName)
        if (result == -1) {
            throw ServerException(
                    String.format("Removing attribute '%s' from file '%s' failed.", attribName, path))
        }
    }
}
